Add default values to the editor help, docs, and generated RST

bojidar-bg opened the pull request:
Also, make spacing of "=" in the editor help a bit more consistent.

To do:
- [x] Add default property values to DocData.
- [x] Export default property values to .xml classes and then to RST.
- [x] Fix or silence these errors produced while getting the default values.
- [x] Add default values for abstract classes like `BaseButton`.
- [x] Add default values to the Properties listing, and not just description in the in-editor help (and maybe in the rst documentation).
- [x] Ensure there are no leftover crashes from the whole process.
- [x] Fix crash at exit from editor/doctool.
- [x] Update docs to remove default values from the descriptions and run doctool (already works, but I don't want to mess up the diff yet). [Done in #30184]

Some screenshots:
![Vector3 docs](https://user-images.githubusercontent.com/5276727/58749391-391dfd80-848e-11e9-9ad6-ec66dea5ac8c.png)
![Button docs](https://user-images.githubusercontent.com/5276727/58749404-6d91b980-848e-11e9-8791-216ab990c675.png)
![GridContainer docs](https://user-images.githubusercontent.com/5276727/58749409-77b3b800-848e-11e9-9da1-93166ef0769b.png)



Closes #16086

bojidar-bg committed:
Add default values to the editor help, docs, and generated RST

Also, make spacing of "=" in the editor help a bit more consistent.
Closes #16086

swarnimarun commented:
This might be the most awesome feature in 3.2 that very few will realize was added.    :D

swarnimarun commented:
Also can't we similarly add this to the Properties section it would be much quicker to be able to see all the default values together? 
I mean the normal Properties section, not Theme Properties.

bojidar-bg commented:
@swarnimarun yeah, going to add it to the normal properties section. Honestly, I was thinking of the same, but wasn't really sure about it...

bojidar-bg commented:
Marked as ready for review. While I haven't updated the documentation yet, I would like to have the other things checked first, so that no unnecessary merge conflicts occur.

akien-mga commented on editor/editor_settings.cpp:
```diff
@@ -1493,6 +1497,15 @@ Ref<ShortCut> ED_SHORTCUT(const String &p_path, const String &p_name, uint32_t p
+		sc->set_name(p_name);
+		sc->set_shortcut(ie);
+		sc->set_meta("original", ie);
+		return sc;
```
Shouldn't it just return `NULL` at the start of the method? What's the use case for `ED_SHORTCUT` when EditorSettings does not exist yet?

bojidar-bg replied:
The use case is certain constructors, such as EditorFileDialog, which use `ED_SHORTCUT` without EditorSettings being instantiated first. I guess returning NULL is valid too, but decided to make it behave as if the shortcut isn't found.

akien-mga commented on main/main.cpp:
```diff
@@ -1372,6 +1372,8 @@ bool Main::start() {
 #ifdef TOOLS_ENABLED
 	if (doc_tool != "") {
 
+		Engine::get_singleton()->set_editor_hint(true); // Needed to instance editor-only classes for their default values
```
Won't this add various editor-only classes to the docs? See e.g. my changes in https://github.com/godotengine/godot/pull/29872/files#diff-7c557c73ffdf760d458e6226f4d70c31R43 where I had to check the editor hint to prevent registering `ResourceImporterOGGVorbis`.

akien-mga replied:
Tested it, and it doesn't seem to add `ResourceImporterOGGVorbis` back.

bojidar-bg replied:
This is more about EditorFileDialog and the like, since ClassDB does not allow instantiating them without the editor hint.

akien-mga commented on modules/gdnative/arvr/arvr_interface_gdnative.cpp:
```diff
@@ -99,13 +98,9 @@ int ARVRInterfaceGDNative::get_capabilities() const {
 
 bool ARVRInterfaceGDNative::get_anchor_detection_is_enabled() const {
-	bool enabled;
+	if (interface == NULL) return false;
```
CC @BastiaanOlij

BastiaanOlij replied:
Fine by me, does the same, my only question is why are we removing the ERR_FAIL_COND_V macro? Did I miss a memo or is this an arbitrary change?

bojidar-bg replied:
This is needed (similar to the one below) in order to not spam the console with errors while getting default values of a class.

akien-mga commented on modules/gdnative/net/multiplayer_peer_gdnative.cpp:
```diff
@@ -68,7 +68,7 @@ void MultiplayerPeerGDNative::set_transfer_mode(TransferMode p_mode) {
 
 NetworkedMultiplayerPeer::TransferMode MultiplayerPeerGDNative::get_transfer_mode() const {
-	ERR_FAIL_COND_V(interface == NULL, TRANSFER_MODE_UNRELIABLE);
+	if (interface == NULL) return TRANSFER_MODE_UNRELIABLE;
```
CC @Faless

Faless replied:
To be honest I don't like removing the ERR_FAILs much, could we add an optional `DEFVAL` override to `ADD_PROPERTY` (`ClassDB::add_property`), which populate the `default_values` used in `ClassDB::class_get_default_property_value`? Would that impact memory or do we load default values anyway? @bojidar-bg 
Otherwise really nice work, it's going to be a great feature! :heart:

bojidar-bg replied:
I guess having a way to specify the default value in `ADD_PROPERTY` would be pretty cool indeed, as it would allow me to not remove the checks. Will look into it..

akien-mga commented on scene/main/scene_tree.cpp:
```diff
@@ -1962,7 +1962,7 @@ bool SceneTree::is_using_font_oversampling() const {
 SceneTree::SceneTree() {
 
-	singleton = this;
+	if (singleton == NULL) singleton = this;
```
So we can no longer instantiate new `SceneTree`s to take over the main loop? Owwww :'(

bojidar-bg replied:
No, previously, instantiating a `SceneTree` and freeing it immediately afterwards resulted in `SceneTree::get_singleton() == NULL`. This change (and the change a few lines down) is intended to keep only the first SceneTree as the singleton tree.

akien-mga commented:
Changes look great overall. I agree that running `doctool` can wait for this to be merged, and be done in a separate PR.

I'm not sure that the way default values are displayed is clear enough for readers. It kind of makes it look like all those have a fixed value instead of a default value. Is this the usual way API docs mention their default values?

akien-mga commented:
I tried the PR locally, I get a segfault running `doctool`:
```
Generating new docs...
WARNING: cleanup: ObjectDB Instances still exist!
   At: core/object.cpp:2111.
pure virtual method called
terminate called without an active exception
Aborted (core dumped)
```

It still works fine though.

akien-mga commented:
> Update docs to remove default values from the descriptions and run doctool (already works, but I don't want to mess up the diff yet).

BTW, I'd suggest to make first one commit that adds the default values everywhere (so just running doctool), and later on another commit that removes all the now-unnecessary default values from descriptions. Would be more readable.

bojidar-bg commented:
> I tried the PR locally, I get a segfault running doctool:

And I get segfaults when running the editor, even without this PR. I think it is related to how default values are cleaned up when they are of object type.

> I'm not sure that the way default values are displayed is clear enough for readers. It kind of makes it look like all those have a fixed value instead of a default value. Is this the usual way API docs mention their default values?

Suggestions welcome. Maybe putting the default in parenthesis?
```
int separation (= 8)
# or, to not have a smiley
int separation (default: 8)
```

Calinou commented:
@bojidar-bg `int separation (default: 8)` looks good to me.

akien-mga commented:
>> I tried the PR locally, I get a segfault running doctool:
>
> And I get segfaults when running the editor, even without this PR. I think it is related to how default values are cleaned up when they are of object type.

I don't get a crash when running `doctool` without this PR though. Here's the `gdb` backtrace running doctool with this PR:

```
pure virtual method called
terminate called without an active exception

Thread 1 "godot-git" received signal SIGABRT, Aborted.
0x000000367ec3ca7a in raise () from /lib64/libc.so.6
(gdb) bt
#0  0x000000367ec3ca7a in raise () from /lib64/libc.so.6
#1  0x000000367ec25524 in abort () from /lib64/libc.so.6
#2  0x000000367fa8d9a3 in __gnu_cxx::__verbose_terminate_handler() [clone .cold.1] () from /lib64/libstdc++.so.6
#3  0x000000367fa938e6 in __cxxabiv1::__terminate(void (*)()) () from /lib64/libstdc++.so.6
#4  0x000000367fa93921 in std::terminate() () from /lib64/libstdc++.so.6
#5  0x000000367fa9463f in __cxa_pure_virtual () from /lib64/libstdc++.so.6
#6  0x0000000003685d29 in ObjectDB::remove_instance (p_object=0x590b5b0) at core/object.cpp:2056
#7  0x0000000003685ab4 in Object::~Object (this=0x590b5b0, __in_chrg=<optimized out>) at core/object.cpp:2014
#8  0x00000000036a54de in Reference::~Reference (this=0x590b5b0, __in_chrg=<optimized out>) at core/reference.cpp:113
#9  0x00000000036b6ee9 in Resource::~Resource (this=0x590b5b0, __in_chrg=<optimized out>) at core/resource.cpp:422
#10 0x0000000002fd6c86 in Curve2D::~Curve2D (this=0x590b5b0, __in_chrg=<optimized out>) at scene/resources/curve.h:150
#11 0x0000000001426ada in memdelete<Reference> (p_class=0x590b5b0) at ./core/os/memory.h:120
#12 0x0000000001425dfa in Ref<Reference>::unref (this=0x61e5098) at ./core/reference.h:282
#13 0x00000000036a4efc in RefPtr::unref (this=0x61e5098) at core/ref_ptr.cpp:82
#14 0x00000000036fb0ed in Variant::clear (this=0x61e5088) at core/variant.cpp:1110
#15 0x00000000013e94d2 in Variant::~Variant (this=0x61e5088, __in_chrg=<optimized out>) at ./core/variant.h:422
#16 0x0000000002f12428 in HashMap<StringName, Variant, HashMapHasherDefault, HashMapComparatorDefault<StringName>, (unsigned char)3, (unsigned char)8>::Pair::~Pair (this=0x61e5080, __in_chrg=<optimized out>)
    at ./core/hash_map.h:61
#17 0x0000000002f143f4 in HashMap<StringName, Variant, HashMapHasherDefault, HashMapComparatorDefault<StringName>, (unsigned char)3, (unsigned char)8>::Element::~Element (this=0x61e5070, 
    __in_chrg=<optimized out>) at ./core/hash_map.h:73
#18 0x0000000002f14423 in memdelete<HashMap<StringName, Variant, HashMapHasherDefault, HashMapComparatorDefault<StringName>, (unsigned char)3, (unsigned char)8>::Element> (p_class=0x61e5070)
    at ./core/os/memory.h:120
#19 0x0000000002f1214a in HashMap<StringName, Variant, HashMapHasherDefault, HashMapComparatorDefault<StringName>, (unsigned char)3, (unsigned char)8>::clear (this=0x6234908) at ./core/hash_map.h:541
#20 0x0000000002f0c996 in HashMap<StringName, Variant, HashMapHasherDefault, HashMapComparatorDefault<StringName>, (unsigned char)3, (unsigned char)8>::~HashMap (this=0x6234908, __in_chrg=<optimized out>)
    at ./core/hash_map.h:602
#21 0x000000000360142e in HashMap<StringName, HashMap<StringName, Variant, HashMapHasherDefault, HashMapComparatorDefault<StringName>, (unsigned char)3, (unsigned char)8>, HashMapHasherDefault, HashMapComparatorD--Type <RET> for more, q to quit, c to continue without paging--
efault<StringName>, (unsigned char)3, (unsigned char)8>::Pair::~Pair (this=0x6234900, __in_chrg=<optimized out>) at ./core/hash_map.h:61
#22 0x000000000360145a in HashMap<StringName, HashMap<StringName, Variant, HashMapHasherDefault, HashMapComparatorDefault<StringName>, (unsigned char)3, (unsigned char)8>, HashMapHasherDefault, HashMapComparatorDefault<StringName>, (unsigned char)3, (unsigned char)8>::Element::~Element (this=0x62348f0, __in_chrg=<optimized out>) at ./core/hash_map.h:73
#23 0x0000000003601489 in memdelete<HashMap<StringName, HashMap<StringName, Variant, HashMapHasherDefault, HashMapComparatorDefault<StringName>, (unsigned char)3, (unsigned char)8>, HashMapHasherDefault, HashMapComparatorDefault<StringName>, (unsigned char)3, (unsigned char)8>::Element> (p_class=0x62348f0) at ./core/os/memory.h:120
#24 0x00000000035fe860 in HashMap<StringName, HashMap<StringName, Variant, HashMapHasherDefault, HashMapComparatorDefault<StringName>, (unsigned char)3, (unsigned char)8>, HashMapHasherDefault, HashMapComparatorDefault<StringName>, (unsigned char)3, (unsigned char)8>::clear (this=0x5047be0 <ClassDB::default_values>) at ./core/hash_map.h:541
#25 0x00000000035fce05 in ClassDB::cleanup () at core/class_db.cpp:1464
#26 0x00000000036a737b in unregister_core_types () at core/register_core_types.cpp:281
#27 0x00000000014243bd in Main::cleanup () at main/main.cpp:2109
#28 0x00000000013e7815 in main (argc=3, argv=0x7fffffffd9e8) at platform/x11/godot_x11.cpp:56
```

I get a similar crash when closing the editor with this PR:
```
pure virtual method called
terminate called without an active exception

Thread 1 "godot-git" received signal SIGABRT, Aborted.
0x000000367ec3ca7a in raise () from /lib64/libc.so.6
Missing separate debuginfos, use: debuginfo-install lib64drm_intel1-2.4.98-1.mga7.x86_64 lib64pciaccess0-0.14-2.mga7.x86_64
(gdb) bt
#0  0x000000367ec3ca7a in raise () from /lib64/libc.so.6
#1  0x000000367ec25524 in abort () from /lib64/libc.so.6
#2  0x000000367fa8d9a3 in __gnu_cxx::__verbose_terminate_handler() [clone .cold.1] () from /lib64/libstdc++.so.6
#3  0x000000367fa938e6 in __cxxabiv1::__terminate(void (*)()) () from /lib64/libstdc++.so.6
#4  0x000000367fa93921 in std::terminate() () from /lib64/libstdc++.so.6
#5  0x000000367fa9463f in __cxa_pure_virtual () from /lib64/libstdc++.so.6
#6  0x0000000003685d29 in ObjectDB::remove_instance (p_object=0x64c53c0) at core/object.cpp:2056
#7  0x0000000003685ab4 in Object::~Object (this=0x64c53c0, __in_chrg=<optimized out>) at core/object.cpp:2014
#8  0x00000000036a54de in Reference::~Reference (this=0x64c53c0, __in_chrg=<optimized out>) at core/reference.cpp:113
#9  0x00000000036b6ee9 in Resource::~Resource (this=0x64c53c0, __in_chrg=<optimized out>) at core/resource.cpp:422
#10 0x0000000002fd6c86 in Curve2D::~Curve2D (this=0x64c53c0, __in_chrg=<optimized out>) at scene/resources/curve.h:150
#11 0x0000000001426ada in memdelete<Reference> (p_class=0x64c53c0) at ./core/os/memory.h:120
#12 0x0000000001425dfa in Ref<Reference>::unref (this=0x6586bd8) at ./core/reference.h:282
#13 0x00000000036a4efc in RefPtr::unref (this=0x6586bd8) at core/ref_ptr.cpp:82
#14 0x00000000036fb0ed in Variant::clear (this=0x6586bc8) at core/variant.cpp:1110
#15 0x00000000013e94d2 in Variant::~Variant (this=0x6586bc8, __in_chrg=<optimized out>) at ./core/variant.h:422
#16 0x0000000002f12428 in HashMap<StringName, Variant, HashMapHasherDefault, HashMapComparatorDefault<StringName>, (unsigned char)3, (unsigned char)8>::Pair::~Pair (this=0x6586bc0, __in_chrg=<optimized out>)
    at ./core/hash_map.h:61
#17 0x0000000002f143f4 in HashMap<StringName, Variant, HashMapHasherDefault, HashMapComparatorDefault<StringName>, (unsigned char)3, (unsigned char)8>::Element::~Element (this=0x6586bb0, 
    __in_chrg=<optimized out>) at ./core/hash_map.h:73
#18 0x0000000002f14423 in memdelete<HashMap<StringName, Variant, HashMapHasherDefault, HashMapComparatorDefault<StringName>, (unsigned char)3, (unsigned char)8>::Element> (p_class=0x6586bb0)
    at ./core/os/memory.h:120
#19 0x0000000002f1214a in HashMap<StringName, Variant, HashMapHasherDefault, HashMapComparatorDefault<StringName>, (unsigned char)3, (unsigned char)8>::clear (this=0x6564bf8) at ./core/hash_map.h:541
#20 0x0000000002f0c996 in HashMap<StringName, Variant, HashMapHasherDefault, HashMapComparatorDefault<StringName>, (unsigned char)3, (unsigned char)8>::~HashMap (this=0x6564bf8, __in_chrg=<optimized out>)
    at ./core/hash_map.h:602
#21 0x000000000360142e in HashMap<StringName, HashMap<StringName, Variant, HashMapHasherDefault, HashMapComparatorDefault<StringName>, (unsigned char)3, (unsigned char)8>, HashMapHasherDefault, HashMapComparatorD--Type <RET> for more, q to quit, c to continue without paging--
efault<StringName>, (unsigned char)3, (unsigned char)8>::Pair::~Pair (this=0x6564bf0, __in_chrg=<optimized out>) at ./core/hash_map.h:61
#22 0x000000000360145a in HashMap<StringName, HashMap<StringName, Variant, HashMapHasherDefault, HashMapComparatorDefault<StringName>, (unsigned char)3, (unsigned char)8>, HashMapHasherDefault, HashMapComparatorDefault<StringName>, (unsigned char)3, (unsigned char)8>::Element::~Element (this=0x6564be0, __in_chrg=<optimized out>) at ./core/hash_map.h:73
#23 0x0000000003601489 in memdelete<HashMap<StringName, HashMap<StringName, Variant, HashMapHasherDefault, HashMapComparatorDefault<StringName>, (unsigned char)3, (unsigned char)8>, HashMapHasherDefault, HashMapComparatorDefault<StringName>, (unsigned char)3, (unsigned char)8>::Element> (p_class=0x6564be0) at ./core/os/memory.h:120
#24 0x00000000035fe860 in HashMap<StringName, HashMap<StringName, Variant, HashMapHasherDefault, HashMapComparatorDefault<StringName>, (unsigned char)3, (unsigned char)8>, HashMapHasherDefault, HashMapComparatorDefault<StringName>, (unsigned char)3, (unsigned char)8>::clear (this=0x5047be0 <ClassDB::default_values>) at ./core/hash_map.h:541
#25 0x00000000035fce05 in ClassDB::cleanup () at core/class_db.cpp:1464
#26 0x00000000036a737b in unregister_core_types () at core/register_core_types.cpp:281
#27 0x00000000014243bd in Main::cleanup () at main/main.cpp:2109
#28 0x00000000013e7815 in main (argc=4, argv=0x7fffffffd9c8) at platform/x11/godot_x11.cpp:56
```

Seems to be the same backtrace.

bojidar-bg commented:
After trying multiple way (`x [= 0.0]`, `x = 0.0`, and `x (default: Color(0, 0, 0, 1))`) to represent default values, I finally chose this one:
![image](https://user-images.githubusercontent.com/5276727/60279458-bf045b80-9909-11e9-9c27-7ea2c40f256f.png)
![image](https://user-images.githubusercontent.com/5276727/60279481-ccb9e100-9909-11e9-9176-4dff652612c2.png)

Also, added `ADD_PROPERTY_DEFAULT` and `ClassDB:set_property_default_value(cl, nm, dv)`, and  reverted the removal of errors from various getters.

akien-mga commented:
Thanks a ton! :tada:
